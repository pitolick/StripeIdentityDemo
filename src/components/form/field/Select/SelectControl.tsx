/**
 * セレクトボックス制御コンポーネント（react-hook-form）
 * @module SelectControl
 */
import React, { forwardRef } from 'react'
import { Select, SelectProps } from './Select'
import { InputBox, InputBoxProps } from '@/components/form/parts'

/**
 * セレクトボックスコンポーネントのプロパティ
 * @memo バリデーションスキーマ（zod）は各入力子コンポーネント内でexportしformコンポーネント内でresolverに指定する
 * @typedef {Object} SelectControlProps
 */
export interface SelectControlProps extends SelectProps, Omit<InputBoxProps, 'children'> {}

/**
 * セレクトボックス制御コンポーネント（react-hook-form）
 * @param {SelectControlProps} props - セレクトボックスコンポーネントのプロパティ
 * @returns {JSX.Element} - セレクトボックスコンポーネント
 */
export const SelectControl = forwardRef<HTMLSelectElement, SelectControlProps>((props, ref) => {
  const { nameId, label, defaultValue, note, required = undefined, options, className, error, contentWidth, ...selectProps } = props
  const classNameProps = className ? className : ''
  // エラー時のclass名
  const errorClass = error ? `bg-red-pale outline outline-red outline-2 -outline-offset-2` : ''
  return (
    <InputBox nameId={nameId} label={label} note={note} required={required} error={error} contentWidth={contentWidth}>
      <Select ref={ref} id={nameId} options={options} defaultValue={defaultValue} className={`${classNameProps} ${errorClass}`} {...selectProps} />
    </InputBox>
  )
})

SelectControl.displayName = 'SelectControl'
