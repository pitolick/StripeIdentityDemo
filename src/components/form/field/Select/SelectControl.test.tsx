import React from 'react'
import { render } from '@testing-library/react'
import { SelectControl, SelectControlProps } from './SelectControl'

describe('SelectControl', () => {
  it('正しくレンダリングされること', () => {
    const props: SelectControlProps = {
      nameId: 'mySelect',
      label: 'マイセレクト',
      defaultValue: 'option1',
      note: { list: ['オプションを選択してください'] },
      required: true,
      options: [
        { value: 'option1', label: 'Option 1' },
        { value: 'option2', label: 'Option 2' },
        { value: 'option3', label: 'Option 3' },
      ],
      className: 'my-select',
      contentWidth: 'full',
      error: { type: 'test' },
    }

    const { container } = render(<SelectControl {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('デフォルト値が正しく設定されること', () => {
    const props: SelectControlProps = {
      nameId: 'mySelect',
      options: [
        { value: 'option1', label: 'Option 1' },
        { value: 'option2', label: 'Option 2' },
        { value: 'option3', label: 'Option 3' },
      ],
    }

    const { container } = render(<SelectControl {...props} />)
    expect(container).toMatchSnapshot()
  })
})
