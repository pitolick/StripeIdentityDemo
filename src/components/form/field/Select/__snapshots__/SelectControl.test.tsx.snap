// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectControl デフォルト値が正しく設定されること 1`] = `
<div>
  <div
    class="false space-y-2 false"
  >
    <div
      class="space-y-2 false"
    >
      <div
        class="flex w-full items-center justify-start gap-x-3"
      >
        <select
          class="  block w-full rounded-md border-gray-500 py-2 pl-3 pr-7 text-sm md:py-3 md:pl-5 md:pr-9 md:text-base"
          id="mySelect"
        >
          <option
            label="Option 1"
            value="option1"
          >
            Option 1
          </option>
          <option
            label="Option 2"
            value="option2"
          >
            Option 2
          </option>
          <option
            label="Option 3"
            value="option3"
          >
            Option 3
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;

exports[`SelectControl 正しくレンダリングされること 1`] = `
<div>
  <div
    class="w-full flex-auto space-y-2 false"
  >
    <div
      class="flex items-center justify-start gap-x-2"
    >
      <span
        class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
      >
        必須
      </span>
      <label
        class="block text-sm font-bold"
        for="mySelect"
      >
        マイセレクト
      </label>
    </div>
    <div
      class="space-y-2 false"
    >
      <div
        class="flex w-full items-center justify-start gap-x-3"
      >
        <select
          class="my-select bg-red-pale outline outline-red outline-2 -outline-offset-2 block w-full rounded-md border-gray-500 py-2 pl-3 pr-7 text-sm md:py-3 md:pl-5 md:pr-9 md:text-base"
          id="mySelect"
        >
          <option
            label="Option 1"
            selected=""
            value="option1"
          >
            Option 1
          </option>
          <option
            label="Option 2"
            value="option2"
          >
            Option 2
          </option>
          <option
            label="Option 3"
            value="option3"
          >
            Option 3
          </option>
        </select>
      </div>
      <ul>
        <li
          class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
        >
          オプションを選択してください
        </li>
      </ul>
      <p
        class="text-sm text-red"
      />
    </div>
  </div>
</div>
`;
